db/components.go

```diff
@@ -0,0 +1,115 @@
+package db
+
+import (
+	"context"
+	"encoding/json"
+
+	"github.com/flanksource/commons/logger"
+	"github.com/flanksource/incident-commander/api"
+	"github.com/flanksource/incident-commander/db/types"
+	"github.com/flanksource/incident-commander/utils"
+	"github.com/google/uuid"
+	"gorm.io/gorm/clause"
+)
+
+func GetTeamsWithComponentSelector() map[uuid.UUID][]api.ComponentSelector {
+	var teams []api.Team
+	var teamComponentMap = make(map[uuid.UUID][]api.ComponentSelector)
+	err := Gorm.Table("teams").Where("spec::jsonb ? 'components';").Find(&teams).Error
+	if err != nil {
+		logger.Errorf("error fetching the teams with componenets: %v", err)
+	}
+	for _, team := range teams {
+		teamSpec := &api.TeamSpec{}
+		teamSpecJson, err := team.Spec.MarshalJSON()
+		if err != nil {
+			logger.Errorf("error marshalling team spec for team: %v", team.ID)
+			continue
+		}
+		if err := json.Unmarshal(teamSpecJson, teamSpec); err != nil {
+			logger.Errorf("error unmarshalling the teamSpec for team: %v", team.ID)
+		}
+		teamComponentMap[team.ID] = teamSpec.Components
+	}
+	return teamComponentMap
+}
+
+func GetComponentsWithSelectors(componentSelectors []api.ComponentSelector) map[string][]uuid.UUID {
+	var selectedComponents = make(map[string][]uuid.UUID)
+	for _, compSelector := range componentSelectors {
+		selectedComponents[utils.GetHash(compSelector)] = getComponentsWithSelector(compSelector)
+	}
+
+	return selectedComponents
+}
+
+func getComponentsWithSelector(selector api.ComponentSelector) []uuid.UUID {
+	sql := "select ID from components where deleted_at is null "
+	var compIds []uuid.UUID
+	args := make(map[string]interface{})
+	if selector.Name != "" {
+		sql += " AND name = :name"
+		args["name"] = selector.Name
+	}
+
+	if selector.Namespace != "" {
+		sql += " AND namespace = :namespace"
+		args["namespace"] = selector.Namespace
+	}
+
+	if selector.Type != "" {
+		sql += " AND type = :type"
+		args["type"] = selector.Type
+	}
+
+	if selector.Labels != nil {
```

Comment:
Could we use chaining here ? https://gorm.io/docs/method_chaining.html

```golang

query = db.Model(Components).Where()

if namespace ! = {
  query = query.Where("namespace  = ?", namespace)
}

if selector != "" {
  query = query.Where("type = "", type)
}
```

and so on, seems cleaner and easier to read

Reply:
Yeah this make more sense